### DISCUSSION TITLE
Do you already know Houdini?

### DISCUSSION THREAD

**Discussion by frederikhors** [jycouet/kitql#70] [Ideas]
A few hours ago I discovered [this amazing project](https://github.com/AlecAivazis/houdini) by @AlecAivazis.

How about joining forces and creating a better library with more brains working on just one thing?

**Comment by jycouet** [comment]
No, I didn't know; thx for the link. _(So you are abandoning KitQL?)_
I had a quick look, and it seems good 👍

I'm very open to `joining forces`. I just have to know a bit more about it 😉
I will investigate a bit further in the coming days to see the different approaches.

Stay tuned

**Reply by frederikhors** [reply]
> So you are abandoning KitQL?

Nope. I'm not using either one yet because I need the Partial results mechanism. I have to understand if I can already manage with Houdini today.

And I don't abandon KitQL because as you can see I immediately linked you to something to learn better about how to solve problems, something that allows us to better open our brains.

**Reply by jycouet** [reply]
I like that :👍

When you say `partial`. It's to go from list to detail, right? (with some fields from the cache that we know already?)

**Reply by frederikhors** [reply]
> When you say `partial`. It's to go from list to detail, right? (with some fields from the cache that we know already?)

Yep. I don't understand how most SPA developers don't consider this small but crucial aspect.

**Reply by jycouet** [reply]
I do have this concern 😅

And it's already working with KitQL. 🥳
Check this: https://svelte.dev/repl/8f2656d0f4594ea6a70eff48a7559ce9?version=3.46.4

When you click on `get Countries`, I simulate a long network call... to get the data.
Before that, I "Patch" my store with all data I know that will come with the updated query.
So I see loading... AND I see the continent because I know it already.

I'm not 100% sure about my "patch" strategy & robustness... That's why I don't communicate so much about it today, but it's definitely something that I want to support fast 😉

**Reply by frederikhors** [reply]
I don't understand that example. I keep seeing Loading message on each click. And what I need is better represented with a click to detail page.

**Reply by jycouet** [reply]
I thought that you would understand the idea behind it.
I removed the "custom" `isLoading` and now if you set the network to 3G you will see the behavior.
=> You already see the country because you patch the store with it then it's querying all the list in the background.
_Today, patch is clearing the cache also to make it simple._

The left side is representing one page right side another page. Anyway, it's a store that will be present globally. so it's the same as a click to a detail page no?

At least, is it going in the right direction? 
(I can do a repo in sveltekit to show you as well)

**Reply by frederikhors** [reply]
I will try in a few hours.

**Reply by frederikhors** [reply]
I studied your code (I did not before). Yes this is one step of a long journey towards the goal. This way of patching is something more similar to Svelte/React-Query (or SWR) of doing things. I do not expect that manually code (the patch() func) to be called if I already have graphql __typenames and IDs. Which is what urql does today with all the problems of urql and with the cons that it is not Svelte native like kitql and Houdini instead.

**Reply by jycouet** [reply]
Today we have this: 
![image](https://user-images.githubusercontent.com/5312607/158490682-2dbc005d-eb21-4f6e-a28a-ea7deb4c33e5.png)
With a manual patch function.

In the last few days, I have been thinking about an automatic patch... But I have a lot of scenarios where it's NOT working. Or where I don't want "half smart" thing.

Let's try something:
```graphql
query posts {
  id
  title
  author
}

query adminQuery {
  posts {
    id
    dateOfLastComment
    author
  }
}

query post {
  id
  title
  description
  dateOfLastComment
}
```

How to build the `partial` of post? 
Should I:
- [ ] Take only `posts.title`
- [ ] Take only `adminQuery.posts.dateOfLastComment`
- [ ] merge `posts.title` & `adminQuery.posts.dateOfLastComment`


Merging is risky as the 2 queries could have been done in a very different timing.
Or I should enforce it ONLY if you use the same fragment. To say: "Yes, this has the same context"

That's why today It's a manual patch, to know what you are doing.
If you have ideas also, let me know.

**Reply by frederikhors** [reply]
> Merging is risky as the 2 queries could have been done in a very different timing.

This can always happen. Not just for `dateOfLastComment`. Even with urql now I have a possible "timing" problem, but by indicating to the user in some way that there is an "underground" query that is checking if the data is correct it is the user himself who "perceives" automatically that the data is possibly stale.

I think we should always merge the data we have using `__typename` and `ID` for this purpose.

**Reply by jycouet** [reply]
What is stale: `title`? `dateOfLastComment`? both? Sometimes one? Sometimes the other one?
Automatic is good in some cases, bad in others. So we have to be able to optIn or optOut.

Don't get me wrong, I want something like this.
That's why I already have it with the path! Ok, it's manual, but you can do ALL strategies you want around 💪
_I'm playing with it on a real big prod app... And see how to make it better by using it._

**Reply by frederikhors** [reply]
> What is stale?

All fields!

> Automatic is good in some cases, bad in others.

Yes, it's true. Since I really like conventions over configurations I often tend to simplify, making mistakes.

> Don't get me wrong

No problem at all!

> Ok, it's manual, but you can do ALL strategies you want around 💪

I'll try in a few hours in a demo project and at the same time I will write some PR docs here.

**Reply by frederikhors** [reply]
I'm not dead! I'm running out of things and then I'll get to work for this.

But now this came to my mind: if we are automatically generating Svelte stores and using patch() methods for the list/detail, I wonder does it still make sense to use GraphQL?

Of course, if it is already used in an existing project, it is necessary and useful to use it also in the client. But if I'm starting a project from scratch, what advantages do I get since I don't need the normalized cache?

Wouldn't it be better to grpc (with grpc-web) which generates typescript types for me too?

What do you think?

**Reply by jycouet** [reply]
GraphQL is not only useful for "normalized cache". 😅

I would go GraphQL 100% all in all the time 🚀🚀🚀
- All the tooling that exists around (Intelisence, ES Lint GraphQL, Modules, Scalar, Stitching, Mesh, ... and more!)
- A BIG community
- 1 engine => Multi-client (website, admin site, mobile, ...)
- Great way to decouple Client / BackEnd / DatabaseS

You try to solve a narrow use case, for sure you can take some shortcuts. But what about the next step?!
I think that GraphQL is the openness that we need 💪

**Reply by jycouet** [reply]
No worries about when you contribute, that's the life of an OSS project 👍

**Reply by frederikhors** [reply]
I tried a few minutes ago for the first time kitql. I need to say it is super exciting to see my queries as stores. Really I'm impressed! It's amazing.

I tried the following code in an existing SvelteKit-Urql project.

```ts
const invoiceForPatch = $KQL_InvoiceList.data?.invoiceList.nodes.find(
		(i) => i.id.toString() === $page.params.id
	);

	KQL_InvoiceById.patch({
		variables: { id: $page.params.id },
		invoiceByID: invoiceForPatch,
	});

	KQL_InvoiceById.query({ variables: { id: $page.params.id } });
```

The result is impressing and I'm really happy. But I cannot understand something.

If I go back to the list page the invoiceById store keeps fetching (I'm using 10 seconds delay on the backend resolvers) and even if I query a new invoiceById store the previous appears until the next (actual) one is fetched.

**Reply by frederikhors** [reply]
The list of my doubts is longer. I'll create a demo app in a few hours.

**Reply by jycouet** [reply]
Yep, patch is:
- not working like this
- not well explained
- will reset all your cache

Don't put variables in patch, but directly expected data result.

Yes, put a demo app up there and I'll show you 👌

**Reply by jycouet** [reply]
After your demo, I'll open an issue to speak about patch. 👍

**Reply by frederikhors** [reply]
Here we are: https://github.com/frederikhors/demo-kitql-sveltekit.

This is my first attempt to a complex (not now) demo! 😄

**Reply by jycouet** [reply]
Thx for doing this.
I already see a few things😉.

1/ The backend delay simulation cannot be put there because you break all the caching mechanisms like that.
2/ Patch will also remove all caching (because today it's a "dummy" patch thing)
3/ in [code].index.svelte you did it only client way, so if you give the URL to someone it will not be SRR (not a big deal, but With KitQL you can compose better the app I think)

I'll open issues to speak more over there.
1/ https://github.com/jycouet/kitql/issues/74
2/ https://github.com/jycouet/kitql/issues/75
3/ I will PR your demo after. (today?)

**Reply by frederikhors** [reply]
1. Yeah it's totally wrong. Also because on destroy i need to clear the setTimeout. I'll update with new delay option too.

**Reply by jycouet** [reply]
You should not update it yet. I'm working on it!

For the patch, I figured out that you don't ALWAYS want to patch from the list! Especially if you have the detail already in the cache... You will patch with "less" data!
I'm doing something about it. Let's see!

**Reply by frederikhors** [reply]
Great!

**Comment by AlecAivazis** [comment]
Just wanted to hop in here and say that I would be very happy to bring in some more brain power into houdini if @jycouet is interested. There's a lot to do and finding a sustainable way to maintain a successful open source library is very tricky with only one person.

Please let me know if there's anything I can explain or help with

**Reply by frederikhors** [reply]
@AlecAivazis I can help too! Can we please find a way to organize a backlog? What are you using right now? Github projects?

**Reply by AlecAivazis** [reply]
Awesome! At the moment, I'm using a [GitHub project](https://github.com/AlecAivazis/houdini/projects/1) to track the set of features I want to get in for the upcoming 1.0 release. Also you both should feel free to reach out to me on discord if you are a member of the Svelte server to chat about anything
